src/isomorphic/children/ReactChildren.js

```diff
@@ -117,7 +117,7 @@ function mapSingleChildIntoContext(bookKeeping, child, childKey) {
         // traverseAllChildren used to do for objects as children
         keyPrefix +
         (
-          mappedChild !== child ?
+          mappedChild !== child && mappedChild.key ?
           escapeUserProvidedKey(mappedChild.key || '') + '/' :
```

Comment:
Haven't thought about this in super detail yet but why not `mappedChild.key !== child.key` as I suggested in chat?